SleepMode docs w/pending config reference (#401)
loft-sh/vcluster-docs: vcluster_versioned_docs/version-0.22.0/configure/vcluster-yaml/experimental/native-sleep-mode.mdx

---
title: Experimental Sleep Mode
sidebar_label: sleepMode
sidebar_position: 7
sidebar_class_name: pro
description: Configure a vCluster to sleep/wake without an agent
---

import ProAdmonition from '../../../_partials/admonitions/pro-admonition.mdx'
import ExperimentalSleepMode from '../../../_partials/config/experimental/sleepMode.mdx'

<ProAdmonition />

:::warning
Native sleep mode is intended for pre-production use cases only, and comes with some limitations and
caveats.
:::

> ### Configuration examples

Reviewer: Could this be H2?

We could do

\## Configuration

\### Examples

\## How it works

\### Detecting activity

\### Ingress considerations

```yaml title="Sleep after 3 hours of inactivity, anything that does not have the label dont=sleep"
experimental:
  sleepMode:
    enabled: true
    autoSleep:

>       afterInactivity: 3h  # Uses Go's Duration with a max unit of hour

Reviewer: Are other values possible here? Like minutes etc? It would be nice to include a callout with link to the Go's Duration fields in :::tip or :::info admonitions.

experimental:
  sleepMode:
    enabled: true
    timezone: America/Denver
    autoSleep:
      schedule: 30 17 * * 5
      wakeup:
        schedule: 0 7 * * 1
```

### How it works

**Sleeping** deletes bare pods and scale down the following resources:
  * Deployments
  * ReplicaSets
  * ReplicationControllers
  * DaemonSets

**Waking** scales those resources back up, unable to restore bare pods that were killed.  To exempt resources from being
put to sleep, either add the annotation `sleepmode.loft.sh/exclude: true` or configure sleep mode with labels to exclude and
add that label to all the workloads you'd like to keep running.

### Detecting activity

To wake a sleeping cluster or to update the last active time, native sleep mode captures the following:

 * Access of cluster resources.  _Think `kubectl get <resource>`_
 * Attempts to contact Ingress endpoints (NGINX only).

:::note
Ingress activity detection is only available for [NGINX](https://github.com/kubernetes/ingress-nginx/tree/main)
ingress controllers, making use of the [mirror-target](https://github.com/kubernetes/ingress-nginx/blob/main/docs/user-guide/nginx-configuration/annotations.md#mirror)
<!-- vale off -->

> annotation.  If you use this annotation, it will be overwritten for the sleep mode to function.

Reviewer: No need to use vale-on/off
```suggestion
annotation. 

:::info 
The annotation It is overriden by vCluster to enable sleep mode. 
:::
```
